DOC: linalg: expand pinv example

j-bowhay opened the pull request:
Currently the example for `pinv` is not super clear what it is showing (unless you already know the properties of generalised inverses). I have expanded the example to give some more motivation for what it is showing and to show some other properties.

dschmitz89 commented:
Looks good to me, just one remark: for people without a strong linar algebra background (like me :P) the biggest difference between the inverse and pseudo-inverse is that the pseudo-inverse is defined also for non-square matrices. Could this be emphasized somewhere in the beginning and in the least squares example?

j-bowhay commented:
> Looks good to me, just one remark: for people without a strong linar algebra background (like me :P) the biggest difference between the inverse and pseudo-inverse is that the pseudo-inverse is defined also for non-square matrices. Could this be emphasized somewhere in the beginning and in the least squares example?

Thanks for having a look @dschmitz89. I have hopefully addressed your comment.

ilayn commented on scipy/linalg/_basic.py:
```diff
@@ -1294,16 +1294,80 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
+    >>> A = rng.standard_normal((9, 6))
+    >>> B = linalg.pinv(A)
+
+    Condition 1:
```
I think we don't need to spell out Condition n here but an inline comment would suffice 

```
>>> np.allclose(A @ B @ A, A)   # Condition 1
True
>>> np.allclose(B @ A @ B, B)  # Condition 2
True 
...
```

We don't need to teach what pinv is. That would be a half-baked exposition anyways.

ilayn commented on scipy/linalg/_basic.py:
```diff
@@ -1294,16 +1294,80 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
+    True
+
+    If ``A`` is invertible then the Moore-Penrose pseudoinverse is exactly the
+    inverse of ``A`` [1]_.
```
This can be a note, we don't need to demonstrate it. Also numerically it will cause some headache if somebody comes in with an ill-conditioned matrix and claims this line is not true. Not a big problem but just nuisance.

ilayn commented on scipy/linalg/_basic.py:
```diff
@@ -1294,16 +1294,80 @@ def pinv(a, atol=None, rtol=None, return_rank=False, check_finite=True,
+
+    If ``A`` is not invertible then the Moore-Penrose pseudoinverse gives the
+    least squares solution to ``Ax = b`` [1]_. In this example, we will try
+    to fit a straight line to the points ``(0,10), (1, 0), (2, 0)``.
```
This one is I think way too specific and does some inter-library coupling. It would be much easier if we can keep this linalg specific and don't go into stats just for the sake of an example. 

```
    If ``A`` is not invertible then the Moore-Penrose pseudoinverse computes the
    ``x`` solution to ``Ax = b`` such that ``||Ax - b||`` is minimized [1]_.
```

sounds good enough to me.

ilayn reviewed (commented):
Thanks @j-bowhay some comments from my side.

ilayn commented:
Looks pretty comprehensive now. Thank you @j-bowhay !
